src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs

```diff
@@ -763,6 +763,78 @@ private void DecodeDllImportAttribute(ref DecodeWellKnownAttributeArguments<Attr
                     preserveSig);
             }
         }
+
+        private void DecodeModuleInitializerAttribute(DecodeWellKnownAttributeArguments<AttributeSyntax, CSharpAttributeData, AttributeLocation> arguments)
+        {
+            Debug.Assert(arguments.AttributeSyntaxOpt is object);
+
+            if (MethodKind != MethodKind.Ordinary)
+            {
+                // Ignore cases where there will already be a warning due to the AttributeTargets.Method usage
+                // restriction on the attribute definition in the framework.
+                if (MethodKind != MethodKind.Constructor && MethodKind != MethodKind.StaticConstructor)
+                {
+                    arguments.Diagnostics.Add(ErrorCode.ERR_ModuleInitializerMethodMustBeOrdinary, arguments.AttributeSyntaxOpt.Location);
+                }
+                return;
+            }
+
+            Debug.Assert(ContainingType is object);
+
+            if (isInaccessible(this) || containingTypes(this).Any(isInaccessible))
+            {
+                var topLevelType = containingTypes(this).Last();
+                arguments.Diagnostics.Add(ErrorCode.ERR_ModuleInitializerMethodMustBeAccessibleOutsideTopLevelType, arguments.AttributeSyntaxOpt.Location, Name, topLevelType.Name);
+            }
+
+            if (!IsStatic)
+            {
+                arguments.Diagnostics.Add(ErrorCode.ERR_ModuleInitializerMethodMustBeStatic, arguments.AttributeSyntaxOpt.Location, Name);
+            }
+
+            if (ParameterCount > 0)
+            {
+                arguments.Diagnostics.Add(ErrorCode.ERR_ModuleInitializerMethodMustNotHaveParameters, arguments.AttributeSyntaxOpt.Location, Name);
+            }
+
+            if (!ReturnsVoid)
+            {
+                arguments.Diagnostics.Add(ErrorCode.ERR_ModuleInitializerMethodMustReturnVoid, arguments.AttributeSyntaxOpt.Location, Name);
+            }
+
+            if (IsGenericMethod)
+            {
+                arguments.Diagnostics.Add(ErrorCode.ERR_ModuleInitializerMethodMustNotBeGeneric, arguments.AttributeSyntaxOpt.Location, Name);
+            }
+
+            if (ContainingType.IsGenericType)
+            {
+                arguments.Diagnostics.Add(ErrorCode.ERR_ModuleInitializerMethodMustNotBeContainedInGenericType, arguments.AttributeSyntaxOpt.Location, Name);
+            }
+
+            DeclaringCompilation.AddModuleInitializerMethod(this);
```

Comment:
> DeclaringCompilation.AddModuleInitializerMethod(this); [](start = 11, length = 55)

Let's not add methods that didn't pass the validation.

Reply:
Okay to check `arguments.Diagnostics.Any()` for this, or is there a better approach?

Reply:
> Okay to check arguments.Diagnostics.Any() for this, or is there a better approach?

In some cases we simply return early if we reported something. Called ```arguments.Diagnostics.Add``` above, return immediately. When we want to report everything we can, we set a flag that we called ```arguments.Diagnostics.Add``` and then check it after all checks are complete. 


---
In reply to: [418196065](https://github.com/dotnet/roslyn/pull/43301#discussion_r418196065) [](ancestors = 418196065)